### ISSUE TITLE
gumbel_softmax is practically random if data is around 0 to 1

### ISSUE THREAD

**Issue by mizimmer90** [pytorch/pytorch#14634]
## 🐛 Bug

The gumbel_softmax function returns seemingly random output when inputs are distributed around 0 to 1. I think this problem arises from the Gaussian noise added to the softmax; if inputs are of similar order of magnitude, the noise overtakes the output. Even using the gumbel_softmax on one-hot vectors will generate a random one-hot output.

This probably isn't a problem if the softmax is used within a training procedure, but I have a pipeline where inputs to the softmax are "near" one-hot encodings. In lieu of having a hacky solution on my end (i.e. multiplying my input data by 100 or adjusting the eps value to a large positive), maybe there could be an option to specify the softmax noise level, or whether to use a purist version of the gumbel_softmax without noise.

### define number of examples and length of each vector of inputs
n_examples = 3
vec_len = 5
### generate a vector of random inputs
rand_vec = np.random.rand(n_examples, vec_len)
### optimal one hot vector as desired from gumbel_softmax
optimal_one_hot_vec = np.eye(vec_len)[np.argmax(rand_vec, axis=1)]
### convert to torch variables
rand_vec_torch = torch.from_numpy(rand_vec).type(torch.FloatTensor)
one_hot_vec_torch = torch.from_numpy(optimal_one_hot_vec).type(torch.FloatTensor)
### print outputs
print(rand_vec)
print(
    "optimal gumbel_softmax: \n",
    one_hot_vec_torch)
print(
    "gumbel_softmax of the random input data: \n",
    F.gumbel_softmax(rand_vec_torch, hard=True))
print(
    "gumbel_softmax of optimal one hot vector: \n",
    F.gumbel_softmax(one_hot_vec_torch, hard=True))

## Expected behavior

In the example above, we expect the gumbel_softmax of the input to match the ideal, although when running this the one-hot outputs of gumbel_softmax are essentially random.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0): 0.4.1
 - OS (e.g., Linux): CentOS Linux 7 (Core)
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): 
 - Python version: 3.6
 - CUDA/cuDNN version: 8.0.61
 - GPU models and configuration: P100
 - Any other relevant information:

**Comment by SsnL** [comment]
Note that one-hot input doesn't represent a distribution with prob. 1 to be a value because there is a **softmax** step! Instead, you should do an input with one finite entry and others being `-inf` to represent a deterministic distn.

```
>>> import torch
>>> x = torch.full((2, 5), -float('inf'))
>>> x
tensor([[-inf, -inf, -inf, -inf, -inf],
        [-inf, -inf, -inf, -inf, -inf]])
>>> x[0,1]=0
>>> x[1,3]=0
>>> x
tensor([[-inf, 0., -inf, -inf, -inf],
        [-inf, -inf, -inf, 0., -inf]])
>>> torch.nn.functional.gumbel_softmax
<function gumbel_softmax at 0x7fef081d6d08>
>>> torch.nn.functional.gumbel_softmax(x)
tensor([[0., 1., 0., 0., 0.],
        [0., 0., 0., 1., 0.]])
>>> torch.nn.functional.gumbel_softmax(x)
tensor([[0., 1., 0., 0., 0.],
        [0., 0., 0., 1., 0.]])
>>> torch.nn.functional.gumbel_softmax(x)
tensor([[0., 1., 0., 0., 0.],
        [0., 0., 0., 1., 0.]])
```

**Comment by TianhaoFu** [comment]
I think you are right @mizimmer90

**Comment by chengstark** [comment]
Same issue here, wish we can get a better explanation or a fix for gumbel to represent the true distribution
